client/db/src/upgrade.rs

```diff
@@ -39,111 +39,122 @@ const CURRENT_VERSION: u32 = 3;
+			UpgradeError::InvalidJustificationBlock => write!(f, "Invalid justification block"),
+			UpgradeError::Io(err) => write!(f, "Io error: {}", err),
 		}
 	}
+}
 
-	update_version(db_path)
+/// Upgrade database to current version.
+pub fn upgrade_db<Block: BlockT>(db_path: &Path, db_type: DatabaseType) -> UpgradeResult<()> {
+	let db_version = current_version(db_path)?;
+	match db_version {
+		0 => return Err(UpgradeError::UnsupportedVersion(db_version)),
+		1 => {
+			migrate_1_to_2::<Block>(db_path, db_type)?;
+			migrate_2_to_3::<Block>(db_path, db_type)?
+		},
+		2 => migrate_2_to_3::<Block>(db_path, db_type)?,
+		CURRENT_VERSION => (),
+		_ => return Err(UpgradeError::FutureDatabaseVersion(db_version)),
+	}
+	update_version(db_path)?;
+	Ok(())
 }
 
 /// Migration from version1 to version2:
 /// 1) the number of columns has changed from 11 to 12;
 /// 2) transactions column is added;
-fn migrate_1_to_2<Block: BlockT>(
-	db_path: &Path,
-	_db_type: DatabaseType,
-) -> sp_blockchain::Result<()> {
-	let db_path = db_path
-		.to_str()
-		.ok_or_else(|| sp_blockchain::Error::Backend("Invalid database path".into()))?;
+fn migrate_1_to_2<Block: BlockT>(db_path: &Path, _db_type: DatabaseType) -> UpgradeResult<()> {
 	let db_cfg = DatabaseConfig::with_columns(V1_NUM_COLUMNS);
-	let db = Database::open(&db_cfg, db_path).map_err(db_err)?;
-	db.add_column().map_err(db_err)
+	let db = Database::open(&db_cfg, db_path)?;
+	db.add_column().map_err(Into::into)
 }
 
 /// Migration from version2 to version3:
 /// - The format of the stored Justification changed to support multiple Justifications.
-fn migrate_2_to_3<Block: BlockT>(
-	db_path: &Path,
-	_db_type: DatabaseType,
-) -> sp_blockchain::Result<()> {
-	let db_path = db_path
-		.to_str()
-		.ok_or_else(|| sp_blockchain::Error::Backend("Invalid database path".into()))?;
+fn migrate_2_to_3<Block: BlockT>(db_path: &Path, _db_type: DatabaseType) -> UpgradeResult<()> {
 	let db_cfg = DatabaseConfig::with_columns(V2_NUM_COLUMNS);
-	let db = Database::open(&db_cfg, db_path).map_err(db_err)?;
+	let db = Database::open(&db_cfg, db_path)?;
 
 	// Get all the keys we need to update
 	let keys: Vec<_> = db.iter(columns::JUSTIFICATIONS).map(|entry| entry.0).collect();
 
 	// Read and update each entry
 	let mut transaction = db.transaction();
 	for key in keys {
-		if let Some(justification) = db.get(columns::JUSTIFICATIONS, &key).map_err(db_err)? {
+		if let Some(justification) = db.get(columns::JUSTIFICATIONS, &key)? {
 			// Tag each justification with the hardcoded ID for GRANDPA to avoid the dependency on
 			// the GRANDPA crate.
 			// NOTE: when storing justifications the previous API would get a `Vec<u8>` and still
 			// call encode on it.
 			let justification = Vec::<u8>::decode(&mut &justification[..])
-				.map_err(|_| sp_blockchain::Error::Backend("Invalid justification blob".into()))?;
+				.map_err(|_| UpgradeError::InvalidJustificationBlock)?;
```

Comment:
The original meaning was that justification blob failed to decode. Not that the block is invalid.